src/Juvix/Data/SHA256.hs

```diff
@@ -4,13 +4,52 @@ import Crypto.Hash.SHA256 qualified as SHA256
 import Data.ByteString.Base16 qualified as Base16
 import Juvix.Prelude
 
-digestText :: Text -> Text
-digestText =
+hashToText :: ByteString -> Text
+hashToText =
   decodeUtf8Lenient
     . Base16.encode
+
+digestText :: Text -> Text
+digestText =
+  hashToText
     . SHA256.hash
     . encodeUtf8
 
--- | Create a HEX encoded, SHA256 digest of the contents of a file.
+-- | Create a HEX encoded, SHA256 digest of the contents of a file
 digestFile :: (Member Files r) => Path Abs File -> Sem r Text
-digestFile = fmap (decodeUtf8Lenient . Base16.encode . SHA256.hash) . readFileBS'
+digestFile = fmap hashToText . digestFileBS
+
+digestFileBS :: (Member Files r) => Path Abs File -> Sem r ByteString
+digestFileBS = fmap SHA256.hash . readFileBS'
+
+data SHA256Builder :: Effect where
+  BuilderDigestFiles :: (Foldable l) => l (Path Abs File) -> SHA256Builder m ()
+
+makeSem ''SHA256Builder
+
+builderDigestFile :: (Members '[SHA256Builder] r) => Path Abs File -> Sem r ()
+builderDigestFile p = builderDigestFiles [p]
+
+runSHA256Builder :: (Members '[Files] r) => Sem (SHA256Builder ': r) a -> Sem r (Text, a)
+runSHA256Builder m = fmap
+  ( first
+      ( hashToText
+          . SHA256.finalize
+      )
+  )
+  $ reinterpret' m (runState SHA256.init)
+  $ \case
+    BuilderDigestFiles f -> do
+      fs <- mapM readFileBS' (toList f)
```

Comment:
There may be an issue with file system path ordering, but some of my assumptions might be faulty:

(I'm assuming that the digest will change if the order of the input files change.)

For the packageId we require that the digest of the package files doesn't change between runs of juvix etc. However we use the file system to return the list of files and calls like this are not guaranteed to be deterministic. (The returned order of paths may change if the metadata of the files change for example).

So maybe we should add an API `digestSortedFiles` which sorts the input first - or just modify `digestFiles` to do the sorting. 

Apologies if you're sorting the input files somewhere, I couldn't find it.

Reply:
that's a very good point. I'll fix  that

Reply:
fixed in 37f30042532561b29082f60065932cb393172eec